corehq/form_processor/backends/sql/update_strategy.py

```diff
@@ -310,3 +333,48 @@ def _apply_form_transaction(self, transaction):
             filtered_updates = [u for u in case_updates if u.id == self.case.case_id]
             for case_update in filtered_updates:
                 self._apply_case_update(case_update, form)
+
+
+def _transaction_sort_key_function(case):
+    def _transaction_cmp(first_transaction, second_transaction):
+        # if the forms aren't submitted by the same user, just default to server dates
+        if first_transaction.user_id != second_transaction.user_id:
+            return cmp(first_transaction.server_date, second_transaction.server_date)
+        else:
+            form_ids = list(case.xform_ids)
+
+            if first_transaction.xform_id and first_transaction.xform_id == second_transaction.xform_id:
+                # short circuit if they are from the same form
+                return cmp(
+                    _type_sort(first_transaction.action_type),
+                    _type_sort(second_transaction.action_type)
+                )
+
+            def _sortkey(transaction):
+                if not transaction.server_date:
+                    raise MissingServerDate()
+
+                def form_cmp(form_id):
+                    return form_ids.index(form_id) if form_id in form_ids else sys.maxsize
+
+                # if the user is the same you should compare with the special logic below
+                # if the user is not the same you should compare just using received_on
+                return (
+                    # this is sneaky - it's designed to use just the date for the
+                    # server time in case the phone submits two forms quickly out of order
+                    transaction.server_date.date(),
```

Comment:
Some small date differences will be ignored here while others will be significant. For example:
```
Difference is ignored:
  server date 1: 2018-08-02T12:09:53.118314Z -> 2018-08-02
  server date 2: 2018-08-02T12:10:02.477660Z -> 2018-08-02

Difference is significant:
  server date 1: 2018-08-02T23:59:53.118314Z -> 2018-08-02
  server date 2: 2018-08-03T00:00:02.477660Z -> 2018-08-03
```

The actual difference between each pair of dates is the same (less than 10s), but there is a different outcome because of the day transition. Since server dates are UTC, this could happen in the middle of the day for some users, which means it's not an unlikely scenario.

Reply:
No, I'm just pointing out an issue. I didn't suggest a fix because I'm not sure of a good fix.

I would not suggest converting to the domain's timezone because users may work at all hours of the day, so that does not solve the issue in a general way, it just makes it less probable.

One option is to use a `cmp` style comparison where the difference between the two dates being compared is measured and if it is less than a given amount then they are considered equal. I was trying to think of a way to do that with a key function, but I'm not sure it's possible.

Reply:
how does this look? 76ea72a